twine/__main__.py

```diff
@@ -25,7 +27,21 @@ def main() -> Any:
     try:
         return cli.dispatch(sys.argv[1:])
     except (exceptions.TwineException, requests.HTTPError) as exc:
-        return "{}: {}".format(exc.__class__.__name__, exc.args[0])
+        parser = argparse.ArgumentParser()
+        parser.add_argument(
+            "--no-color", default=False, required=False, action="store_true",
+        )
```

Comment:
I think I understand why this is defined here *and* in `settings.py`, but the duplication feels a little awkward.

Also, the current implementation means it only works for `twine upload`. So, even though `twine check` can report errors, we can't disable the color:

<img width="657" alt="Screen Shot 2020-06-07 at 8 19 11 AM" src="https://user-images.githubusercontent.com/1326704/83968475-96b20a80-a897-11ea-871b-2c97fc82a867.png">

I don't have a solid recommendation at this point, since the `parser` instances are local to each command's `main()` function. I'm guessing that could be pulled out into a module attribute, maybe in each command, or maybe once in `cli.py`. And, maybe all of the exception handling/formatting could move to `cli.dispatch`. However, those ideas feel like a significant restructuring, so I'd want to get other maintainers' feedback before proceeding.

Reply:
This is also a problem for the `--verbose` argument, where it only works for `twine upload` and not for `twine check`, so getting these arguments to work for all twine commands will require another PR dedicated towards the restructuring you mentioned. I'm imagining something like pip's [`logging.py`](https://github.com/pypa/pip/blob/e060970d51c5946beac8447eb95585d83019582d/src/pip/_internal/utils/logging.py#L174), but @di might have more to say on this.

Reply:
We need to refactor the commands into something with a unified (internal) API like `TwineCommand`, so we can set flags across all commands. I think that's out of scope for this PR, though.

Reply:
Given that the current implementation means `--no-color` doesn't work for `twine check`, I'm not ready to approve this. I agree that #659 is out of scope, so I think it's worth considering alternatives.

In addition to my previous suggestion for making `parser` a module attribute, another option that comes to mind is a utility function that adds common options to a `parser` instance, e.g.:

```python
def main(args: List[str]) -> bool:
    parser = argparse.ArgumentParser(prog="twine check")
    cli.add_common_options(parser)
```

> This is also a problem for the `--verbose` argument

Also, FWIW, I think `--verbose` is different, because it's behavior will vary depending on the command. For example, `check` might not need a `--verbose` option.

Reply:
The last commit I made fixes the issue with `twine check` not being able to handle options defined in `settings.py`, meaning the `--no-color` option (and options like `--verbose` or anything else that needs to be added in the future) now works for all commands.

I like the idea of a utility function that would do all this; and to your point that the parser logic shouldn't be `__main__.py`, it would also be nice if `settings.Settings` was a singleton object whose parameters (e.g. `verbose`, `no_color`, `disable_progress_bar`) are accessible from anywhere, but that might just solve the issue of having to re-parse the arguments in `__main__.py`. I'm also not sure how common singletons are in Python or if that's best practice for situations like these.